lib/gst/player/gstplayer.c

```diff
@@ -1121,6 +1167,464 @@ element_cb (GstBus * bus, GstMessage * msg, gpointer user_data)
+}
+
+static void
+emit_media_updated_signal (GstPlayer *self, GstPlayerMediaInfo *info,
+  guint signal)
+{
+  if (self->priv->dispatch_to_main_context) {
+    MediaInfoUpdatedSignalData *data = g_slice_new
+                                        (MediaInfoUpdatedSignalData);
+
+    data->player = self;
+    g_mutex_lock (&self->priv->lock);
+    if (self->priv->media_info)
+      g_object_unref (self->priv->media_info);
+    self->priv->media_info = info;
+    data->info   = info;
+    g_mutex_unlock (&self->priv->lock);
+    data->signal = signal;
+
+    g_main_context_invoke_full (self->priv->application_context,
+                    G_PRIORITY_DEFAULT,
+                    media_info_updated_dispatch,
+                    data,
+                    (GDestroyNotify) free_media_info_updated_signal_data);
+  } else {
+    g_signal_emit (self, signals[signal], 0);
+  }
+}
+
+static GstCaps*
+get_caps (GstPlayer *self, gint stream_index, GType type)
+{
+  GstPad  *pad = NULL;
+  GstCaps *caps = NULL;
+
+  if (type == GST_TYPE_PLAYER_VIDEO_INFO)
+    g_signal_emit_by_name (G_OBJECT (self->priv->playbin),
+                          "get-video-pad", stream_index, &pad);
+
+  if (type == GST_TYPE_PLAYER_AUDIO_INFO)
+    g_signal_emit_by_name (G_OBJECT (self->priv->playbin),
+                          "get-audio-pad", stream_index, &pad);
+
+  if (type == GST_TYPE_PLAYER_SUBTITLE_INFO)
+    g_signal_emit_by_name (G_OBJECT (self->priv->playbin),
+                          "get-text-pad", stream_index, &pad);
+
+  if (pad)
+    caps = gst_pad_get_current_caps (pad);
+
+  return caps;
+}
+
+static void
+gst_player_subtitle_info_update (GstPlayer *self,
+  GstPlayerStreamInfo *stream_info)
+{
+  GstPlayerSubtitleInfo *info = (GstPlayerSubtitleInfo*) stream_info;
+
+  if (stream_info->tags) {
+    gchar *lang_code = NULL;
+
+    if (gst_tag_list_get_string (stream_info->tags,
+          GST_TAG_LANGUAGE_CODE, &lang_code)) {
+      info->language = g_strdup (gst_tag_get_language_name (lang_code));
+      g_free (lang_code);
+    }
+    else {
+      gst_tag_list_get_string (stream_info->tags,
+          GST_TAG_LANGUAGE_CODE, &info->language);
+    }
+  }
+
+  GST_DEBUG_OBJECT (self, "language=%s", info->language);
+}
+
+static void
+gst_player_video_info_update (GstPlayer *self,
+  GstPlayerStreamInfo *stream_info)
+{
+  GstPlayerVideoInfo *info = (GstPlayerVideoInfo*) stream_info;
+
+  if (stream_info->caps) {
+    GstStructure *s;
+
+    s = gst_caps_get_structure (stream_info->caps, 0);
+    if (s) {
+      gint width, height;
+      gint fps_n, fps_d;
+      gint par_n, par_d;
+
+      if (gst_structure_get_int (s, "width", &width))
+        info->width = width;
+
+      if (gst_structure_get_int (s, "height", &height))
+        info->height = height;
+
+      if (gst_structure_get_fraction (s, "framerate", &fps_n, &fps_d)) {
+        info->framerate_num = fps_n;
+        info->framerate_denom = fps_d;
```

Comment:
one question on this section, since we are initializing <br>
width=-1, height=-1, fps_n=0, fps_d=1 and par_n=par_d=1<br>
in our class_init section hence else part is not required. If() fails then we are not setting that field hence the field contains the default.